lib/repl.js

```diff
@@ -374,6 +374,18 @@ function REPLServer(prompt,
     };
   }
 
+  Error.prepareStackTrace = function prepareStackTrace(error, structuredStack) {
```

Comment:
hmmm... I do not think it is a good idea to be extending `Error` in this way. Why is the existing `Error.captureStackTrace()` not adequate?

Reply:
@jasnell [`Error.captureStackTrace()`](https://github.com/v8/v8/wiki/Stack-Trace-API#stack-trace-collection-for-custom-exceptions) captures the stack at the point where it is called. This would rarely be the stack a user is looking for, since that call would happen in the internals of `REPLServer`, after a call to `vm.runInContext()`. I think since it would be after the failed completion of an asynchronous execution, the user stack would be gone. 

If I understand the v8 Error API documentation, [`Error.prepareStackTrace()`](https://github.com/v8/v8/wiki/Stack-Trace-API#customizing-stack-traces) is the recommended way to accomplish what I am trying to do.

Reply:
Ok. Lemme stew on this a bit more then :-)

Reply:
I too had the feeling that doing this is somewhat wrong but by thinking about it further I agree with @lance that this is probably the right thing to do for the repl.

Reply:
@lance User may override `Error.prepareStackTrace` in their code.
I suggest something like 👇  
```js
const prepareStackTrace = fn => (error, frames) => { 
	// trim stack frames...
	let trimmedFrames = frames;
	if (fn) {
		// pass trimmed stack frame
		return fn(error, trimmedFrames);
	}

	return trimmedFrames;
};

self.outputStream._write = self.outputStream.write;
self.outputStream.write = (d) => {
	let pstrace = Error.prepareStackTrace;
	Error.prepareStackTrace = prepareStackTrace(pstrace);
	self.outputStream._write(d);
	Error.prepareStackTrace = pstrace;
};
```

Reply:
@princejwesley I thought about that as well but that can be seen as a advantage and as a disadvantage at the same time. Some users might want to know where the repl lines come from (probably only node-core and we could work around it but it is still something to consider). Therefore I did not complain about that.
So do we really want to prevent the user from being able to override the function? If he does that in the first place, the user can also filter the internal lines if he cares to do so.

Reply:
@BridgeAR 
May be setting `process.stdout.write = process.stdout._write` would disable the trimming feature(because we are binding the original function(`outputstream.write`) to same `outputstream` object  instead of keeping it local to context).

Adding comment is sufficient, right? 
Advantages:
- User's choice is preserved.
- repl stack is not leaked.

Reply:
@BridgeAR  we are not preventing user's `printStackTrace` function, Just decorating it with our `printStackTrace` which will trim the stack frames and pass it to user defined `printStackTrace`.

Reply:
@princejwesley @BridgeAR I am a little conflicted about this. Before getting into that though, the example here confuses me a little. What happens when `self.outputStream._write(d);` throws? Wouldn't `Error.prepareStackTrace` never get restored?

Reply:
@lance wrap it with try/finally block. if `self.outputstream._write(d)` throws error, we can't use repl at all.